src/utils/calculateNumbersOfPages.js

```diff
@@ -0,0 +1,54 @@
+//  1 = 1, 2, ... 11, 12
+// 2 = 1, 2, 3, 4, ..., 11, 12
+// 3 = 1, 2, 3, 4, ..., 11, 12
+// 4 = 1, 2, 3, 4, 5, ..., 11, 12
+// 5 = 1, 2, ..., 4, 5, 6, ..., 11, 12
+// ...
+// 9 = 1, 2, ..., 8, 9, 10, 11, 12
+// 10 = 1, 2, ..., 9, 10, 11, 12
+// 11 = 1, 2, ..., 10, 11, 12
+// 12 = 1, 2, ..., 11, 12
+// 99 = 1, 2, ..., 98,99,100
+//100 = 1,2, ... 99, 100
+
+export function calculateNumbersOfPages(totalPages, currentPageState) {
+  const numbersOfPages = [];
+  for (let i = 1; i <= totalPages; i++) {
+    const currentPage = {
+      pageNumber: null,
+      isActive: false,
+    };
+    let doPush = true;
+
+    if (totalPages < 6) {
+      currentPage.pageNumber = i;
+      currentPage.isActive = true;
+    } else {
+      if (
+        (((currentPageState === 1 && i > 2) ||
+          (currentPageState < 4 && i >= 5) ||
+          (currentPageState === 4 && i >= 6) ||
+          (currentPageState >= 5 &&
+            currentPageState < 9 &&
+            i > 2 &&
+            (i < currentPageState - 1 || i > currentPageState + 1))) &&
+          i < totalPages - 1) ||
+        (currentPageState >= 9 && i > 2 && i < currentPageState - 1)
+      ) {
+        const previousPage = numbersOfPages[numbersOfPages.length - 1];
+        doPush = !previousPage || previousPage.isActive;
+
+        currentPage.pageNumber = null;
+        currentPage.isActive = false;
+      } else {
+        currentPage.pageNumber = i;
+        currentPage.isActive = true;
+      }
+    }
+
+    if (doPush) {
+      numbersOfPages.push(currentPage);
+    }
+  }
+  return numbersOfPages;
+}
```

Comment:
This doesn't really work right. For 20 total pages and active page 12, it generates a a following range:

```js
[
  { pageNumber: 1, isActive: true },
  { pageNumber: 2, isActive: true },
  { pageNumber: null, isActive: false },
  { pageNumber: 11, isActive: true },
  { pageNumber: 12, isActive: true },
  { pageNumber: 13, isActive: true },
  { pageNumber: 14, isActive: true },
  { pageNumber: 15, isActive: true },
  { pageNumber: 16, isActive: true },
  { pageNumber: 17, isActive: true },
  { pageNumber: 18, isActive: true },
  { pageNumber: 19, isActive: true },
  { pageNumber: 20, isActive: true }
]
```

while the expected result is as follows:

```js
[
  { isActive: false, pageNumber: 1 },
  { isActive: false, pageNumber: 2 },
  { isActive: false, pageNumber: null },
  { isActive: false, pageNumber: 11 },
  { isActive: true, pageNumber: 12 },
  { isActive: false, pageNumber: 13 },
  { isActive: false, pageNumber: null },
  { isActive: false, pageNumber: 19 },
  { isActive: false, pageNumber: 20 }
]
```

Perhaps your unit test could use some more cases as well :) 

Also this function could use some readability treatment. Perhaps you could utilize more smaller functions? For example, within your `calculateNumbersOfPages` you could create a `createPaginationRange` function that would return you an array of values, for example:

```js
function calculateNumbersOfPages(totalPages, currentPageState) {
    // Creates a range of numbered pages
    const createRange = (from, to) => {
        const pages = to - (from - 1);
        return new Array(pages).fill(null).map((_, index) => {
            const page = x; // where x is the calculated valid page based on `from` and `to`
            return {
                isActive: page === currentPageState,
                pageNumber: page,
            };
        });
    };

    // Creates an empty pagination item with no page number
    const createEmptyItem = () => ({ isActive: false, pageNumber: null });

    // then you can simply return valid configuration of pages based on different conditions, like:
    if (totalPages < 5) {
       return createRange(1, totalPages);
    }


    if (totalPages > 5 && totalPages <= 9) {
        return [
            ...createRange(1, 2),
            createEmptyItem(),
            ...createRange(totalPages - 1, totalPages),
        ];
    }

    // ...... etc etc ......
}
```